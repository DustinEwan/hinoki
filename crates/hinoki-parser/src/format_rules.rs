use crate::Rule;
use pest::error::Error;

pub fn format_parse_error(err: Error<Rule>) -> Error<Rule> {
    err.renamed_rules(|rule| match *rule {
        Rule::EOI => "EOF".to_owned(),
        Rule::file => "import, struct, enum, fn, or EOF".to_owned(),

        // Include other rules above that aren't straightforward when hitting parsing errors!
        other => format!("{:?}", other),
    })
}
