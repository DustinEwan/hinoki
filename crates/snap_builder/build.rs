use std::{
    env,
    fs::{self, DirEntry},
    path::Path,
};

const SCRIPT_PATH: &str = "../hinoki-parser/testdata/snapshots";

fn main() {
    println!("cargo:rerun-if-changed={}", SCRIPT_PATH);
    let mut test_list_string = String::from("static TEST_FILES: [&str;");

    println!("{}", env::current_dir().unwrap().to_str().unwrap());
    let dir = fs::read_dir(SCRIPT_PATH);

    let mut filenames: Vec<String> = vec![];

    match dir {
        Ok(files) => {
            let files = files.flatten().collect::<Vec<DirEntry>>();
            let num_files = files.len();
            test_list_string.push_str(&format!("{}] = [\n", num_files).to_owned());

            files.into_iter().for_each(|file| {
                let filename = file.file_name();
                let filename = filename.to_str().unwrap();
                let filename = filename.replace(".hi", "");

                println!("Found file: {}", filename);

                filenames.push(format!("\t\"{}\"", filename).to_owned());
            });
        }
        Err(e) => panic!("{}", e),
    };

    test_list_string.push_str(&filenames.join(",\n"));
    test_list_string.push_str("\n];\n");

    println!("{}", test_list_string);

    let out_dir = env::var("OUT_DIR").unwrap();
    let dest_path = Path::new(&out_dir).join("testfiles.rs");
    let _ = fs::write(dest_path, test_list_string);
}
