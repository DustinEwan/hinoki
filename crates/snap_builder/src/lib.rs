extern crate proc_macro;
use proc_macro::TokenStream;

include!(concat!(env!("OUT_DIR"), "/testfiles.rs"));

#[proc_macro]
pub fn build_snaps(_item: TokenStream) -> TokenStream {
    TEST_FILES
        .iter()
        .map(|f| format!("snap!({});", f))
        .collect::<Vec<String>>()
        .join("\n")
        .parse()
        .unwrap()
}
