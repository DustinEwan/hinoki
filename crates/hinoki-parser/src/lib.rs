mod format_rules;

use once_cell::sync::OnceCell;
use tracing_subscriber::{fmt::FormattedFields, util::SubscriberInitExt};

use pest::Parser;
use pest::{error::Error, pratt_parser::PrattParser};
use pest_derive::Parser;

use crate::format_rules::format_parse_error;

#[derive(Parser)]
#[grammar = "./hinoki.pest"]
struct HinokiParser;

// lazy_static::lazy_static! {
//     static ref PRATT_PARSER: PrattParser<Rule> = {
//         use pest::pratt_parser::{Assoc::*, Op};
//         use Rule::*;
//
//         // Precedence is defined lowest to highest
//         PrattParser::new()
//             .op(Op::infix(xor, Left))
//             .op(Op::infix(bitwise_or, Left))
//             .op(Op::infix(bitwise_and, Left))
//             .op(Op::infix(not, Left))
//             .op(Op::infix(less_than, Left) | Op::infix(greater_than, Left) | Op::infix(equal, Left) | Op::infix(less_than_or_equal, Left) | Op::infix(greater_than_or_equal, Left) | Op::infix(not_equal, Left) )
//             .op(Op::infix(shift_left, Left) | Op::infix(shift_right, Left))
//             .op(Op::infix(add, Left) | Op::infix(subtract, Left))
//             .op(Op::infix(modulus, Left))
//             .op(Op::infix(multiply, Left) | Op::infix(divide, Left))
//             .op(Op::infix(power, Left))
//     };
// }

pub struct Program {
    pub instructions: Vec<Instruction>,
}

pub enum Instruction {
    Statement(Statement),
    Expression,
}

pub enum Statement {
    Declaration(DeclarationStmt),
    Assignment(AssignmentStmt),
    Return(ReturnStmt),
    Expression(Expression),
}

pub struct DeclarationStmt {}
pub struct AssignmentStmt {}
pub struct ReturnStmt {}

pub enum Expression {
    Integer(i64),
    Float(f64),
    Binary(BinaryExpr),
}

pub struct BinaryExpr {
    lhs: Box<Expression>,
    op: Op,
    rhs: Box<Expression>,
}

pub enum Op {
    Add,
    Subtract,
    Multiply,
    Divide,
    Mod,
    Power,

    ShiftLeft,
    ShiftRight,
    BitwiseAnd,
    BitwiseOr,
    Xor,
    Not,

    LessThan,
    GreaterThan,
    Equal,
    LessThanOrEqual,
    GreaterThanOrEqual,
    NotEqual,
}

// pub fn parse_expr(pairs: Pairs<Rule>) -> Expression {
//     PRATT_PARSER
//         .map_primary(|primary| match primary.as_rule() {
//             Rule::
//         })
// }

fn snapshot_parsing(input: &str) -> String {
    let file = HinokiParser::parse(Rule::file, input);

    match file {
        Ok(mut file) => format!("{:#?}", file.next().unwrap()),
        Err(e) => {
            let formatted_err = format_parse_error(e.clone());
            println!("{}", e);
            format!("{:#?}\n\n{}", e, formatted_err)
        }
    }
}

pub fn setup_trace() {
    static INSTANCE: OnceCell<()> = OnceCell::new();
    INSTANCE.get_or_init(|| {
        tracing_subscriber::fmt()
            .with_max_level(tracing::Level::TRACE)
            .without_time()
            .with_target(false)
            .finish()
            .init();
    });
}

#[cfg(test)]
mod test {
    use snap_builder::build_snaps;

    use crate::*;

    macro_rules! snap {
        ($name: tt) => {
            #[test]
            fn $name() {
                setup_trace();

                let contents =
                    include_str!(concat!("../testdata/snapshots/", stringify!($name), ".hi"));
                let mut settings = insta::Settings::clone_current();

                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_parsing(contents));
                });
            }
        };
    }

    build_snaps!();
    // snap!(test_declaration, "simple_arithmetic.hi");
}
