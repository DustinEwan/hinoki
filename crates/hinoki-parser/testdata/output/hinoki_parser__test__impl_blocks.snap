---
source: crates/hinoki-parser/src/lib.rs
expression: snapshot_parsing(contents)
---
Pair {
    rule: file,
    span: Span {
        str: "\nimpl SomeStruct {\n\n\tfn identity(self): Self {\n\t\tself\n\t}\n\n\tpub fn add(self, a: int, b: int): int {\n\t\ta + b\n\t}\n\n\tfn increase_count(mut self, step: int) {\n\t\tset self.count += step;\n\t}\n\n\tpub fn increase_count(mut self, step: int) {\n\t\tset self.count += step;\n\t}\n\n}\n",
        start: 0,
        end: 261,
    },
    inner: [
        Pair {
            rule: impl_definition,
            span: Span {
                str: "impl SomeStruct {\n\n\tfn identity(self): Self {\n\t\tself\n\t}\n\n\tpub fn add(self, a: int, b: int): int {\n\t\ta + b\n\t}\n\n\tfn increase_count(mut self, step: int) {\n\t\tset self.count += step;\n\t}\n\n\tpub fn increase_count(mut self, step: int) {\n\t\tset self.count += step;\n\t}\n\n}",
                start: 1,
                end: 260,
            },
            inner: [
                Pair {
                    rule: impl,
                    span: Span {
                        str: "impl",
                        start: 1,
                        end: 5,
                    },
                    inner: [],
                },
                Pair {
                    rule: ident,
                    span: Span {
                        str: "SomeStruct",
                        start: 6,
                        end: 16,
                    },
                    inner: [],
                },
                Pair {
                    rule: impl_block,
                    span: Span {
                        str: "{\n\n\tfn identity(self): Self {\n\t\tself\n\t}\n\n\tpub fn add(self, a: int, b: int): int {\n\t\ta + b\n\t}\n\n\tfn increase_count(mut self, step: int) {\n\t\tset self.count += step;\n\t}\n\n\tpub fn increase_count(mut self, step: int) {\n\t\tset self.count += step;\n\t}\n\n}",
                        start: 17,
                        end: 260,
                    },
                    inner: [
                        Pair {
                            rule: impl_function_definition,
                            span: Span {
                                str: "fn identity(self): Self {\n\t\tself\n\t}",
                                start: 21,
                                end: 56,
                            },
                            inner: [
                                Pair {
                                    rule: fn,
                                    span: Span {
                                        str: "fn",
                                        start: 21,
                                        end: 23,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "identity",
                                        start: 24,
                                        end: 32,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: impl_function_parameters_list,
                                    span: Span {
                                        str: "self",
                                        start: 33,
                                        end: 37,
                                    },
                                    inner: [
                                        Pair {
                                            rule: self_param,
                                            span: Span {
                                                str: "self",
                                                start: 33,
                                                end: 37,
                                            },
                                            inner: [],
                                        },
                                    ],
                                },
                                Pair {
                                    rule: self_type,
                                    span: Span {
                                        str: "Self",
                                        start: 40,
                                        end: 44,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: block_expr,
                                    span: Span {
                                        str: "{\n\t\tself\n\t}",
                                        start: 45,
                                        end: 56,
                                    },
                                    inner: [
                                        Pair {
                                            rule: expr,
                                            span: Span {
                                                str: "self\n\t",
                                                start: 49,
                                                end: 55,
                                            },
                                            inner: [
                                                Pair {
                                                    rule: term,
                                                    span: Span {
                                                        str: "self",
                                                        start: 49,
                                                        end: 53,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: ident,
                                                            span: Span {
                                                                str: "self",
                                                                start: 49,
                                                                end: 53,
                                                            },
                                                            inner: [],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                        Pair {
                            rule: impl_function_definition,
                            span: Span {
                                str: "pub fn add(self, a: int, b: int): int {\n\t\ta + b\n\t}",
                                start: 59,
                                end: 109,
                            },
                            inner: [
                                Pair {
                                    rule: pub,
                                    span: Span {
                                        str: "pub",
                                        start: 59,
                                        end: 62,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: fn,
                                    span: Span {
                                        str: "fn",
                                        start: 63,
                                        end: 65,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "add",
                                        start: 66,
                                        end: 69,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: impl_function_parameters_list,
                                    span: Span {
                                        str: "self, a: int, b: int",
                                        start: 70,
                                        end: 90,
                                    },
                                    inner: [
                                        Pair {
                                            rule: self_param,
                                            span: Span {
                                                str: "self",
                                                start: 70,
                                                end: 74,
                                            },
                                            inner: [],
                                        },
                                        Pair {
                                            rule: function_parameter,
                                            span: Span {
                                                str: "a: int",
                                                start: 76,
                                                end: 82,
                                            },
                                            inner: [
                                                Pair {
                                                    rule: ident,
                                                    span: Span {
                                                        str: "a",
                                                        start: 76,
                                                        end: 77,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: type,
                                                    span: Span {
                                                        str: "int",
                                                        start: 79,
                                                        end: 82,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: ident,
                                                            span: Span {
                                                                str: "int",
                                                                start: 79,
                                                                end: 82,
                                                            },
                                                            inner: [],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                        Pair {
                                            rule: function_parameter,
                                            span: Span {
                                                str: "b: int",
                                                start: 84,
                                                end: 90,
                                            },
                                            inner: [
                                                Pair {
                                                    rule: ident,
                                                    span: Span {
                                                        str: "b",
                                                        start: 84,
                                                        end: 85,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: type,
                                                    span: Span {
                                                        str: "int",
                                                        start: 87,
                                                        end: 90,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: ident,
                                                            span: Span {
                                                                str: "int",
                                                                start: 87,
                                                                end: 90,
                                                            },
                                                            inner: [],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                },
                                Pair {
                                    rule: type,
                                    span: Span {
                                        str: "int",
                                        start: 93,
                                        end: 96,
                                    },
                                    inner: [
                                        Pair {
                                            rule: ident,
                                            span: Span {
                                                str: "int",
                                                start: 93,
                                                end: 96,
                                            },
                                            inner: [],
                                        },
                                    ],
                                },
                                Pair {
                                    rule: block_expr,
                                    span: Span {
                                        str: "{\n\t\ta + b\n\t}",
                                        start: 97,
                                        end: 109,
                                    },
                                    inner: [
                                        Pair {
                                            rule: expr,
                                            span: Span {
                                                str: "a + b\n\t",
                                                start: 101,
                                                end: 108,
                                            },
                                            inner: [
                                                Pair {
                                                    rule: term,
                                                    span: Span {
                                                        str: "a",
                                                        start: 101,
                                                        end: 102,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: ident,
                                                            span: Span {
                                                                str: "a",
                                                                start: 101,
                                                                end: 102,
                                                            },
                                                            inner: [],
                                                        },
                                                    ],
                                                },
                                                Pair {
                                                    rule: add,
                                                    span: Span {
                                                        str: "+",
                                                        start: 103,
                                                        end: 104,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: term,
                                                    span: Span {
                                                        str: "b",
                                                        start: 105,
                                                        end: 106,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: ident,
                                                            span: Span {
                                                                str: "b",
                                                                start: 105,
                                                                end: 106,
                                                            },
                                                            inner: [],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                        Pair {
                            rule: impl_function_definition,
                            span: Span {
                                str: "fn increase_count(mut self, step: int) {\n\t\tset self.count += step;\n\t}",
                                start: 112,
                                end: 181,
                            },
                            inner: [
                                Pair {
                                    rule: fn,
                                    span: Span {
                                        str: "fn",
                                        start: 112,
                                        end: 114,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "increase_count",
                                        start: 115,
                                        end: 129,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: impl_function_parameters_list,
                                    span: Span {
                                        str: "mut self, step: int",
                                        start: 130,
                                        end: 149,
                                    },
                                    inner: [
                                        Pair {
                                            rule: mutable,
                                            span: Span {
                                                str: "mut",
                                                start: 130,
                                                end: 133,
                                            },
                                            inner: [],
                                        },
                                        Pair {
                                            rule: self_param,
                                            span: Span {
                                                str: "self",
                                                start: 134,
                                                end: 138,
                                            },
                                            inner: [],
                                        },
                                        Pair {
                                            rule: function_parameter,
                                            span: Span {
                                                str: "step: int",
                                                start: 140,
                                                end: 149,
                                            },
                                            inner: [
                                                Pair {
                                                    rule: ident,
                                                    span: Span {
                                                        str: "step",
                                                        start: 140,
                                                        end: 144,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: type,
                                                    span: Span {
                                                        str: "int",
                                                        start: 146,
                                                        end: 149,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: ident,
                                                            span: Span {
                                                                str: "int",
                                                                start: 146,
                                                                end: 149,
                                                            },
                                                            inner: [],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                },
                                Pair {
                                    rule: block_expr,
                                    span: Span {
                                        str: "{\n\t\tset self.count += step;\n\t}",
                                        start: 151,
                                        end: 181,
                                    },
                                    inner: [
                                        Pair {
                                            rule: mutate_stmt,
                                            span: Span {
                                                str: "set self.count += step;",
                                                start: 155,
                                                end: 178,
                                            },
                                            inner: [
                                                Pair {
                                                    rule: set,
                                                    span: Span {
                                                        str: "set",
                                                        start: 155,
                                                        end: 158,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: assignment_accessor,
                                                    span: Span {
                                                        str: "self.count",
                                                        start: 159,
                                                        end: 169,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: increment_assign,
                                                    span: Span {
                                                        str: "+=",
                                                        start: 170,
                                                        end: 172,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: expr,
                                                    span: Span {
                                                        str: "step",
                                                        start: 173,
                                                        end: 177,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: term,
                                                            span: Span {
                                                                str: "step",
                                                                start: 173,
                                                                end: 177,
                                                            },
                                                            inner: [
                                                                Pair {
                                                                    rule: ident,
                                                                    span: Span {
                                                                        str: "step",
                                                                        start: 173,
                                                                        end: 177,
                                                                    },
                                                                    inner: [],
                                                                },
                                                            ],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                        Pair {
                            rule: impl_function_definition,
                            span: Span {
                                str: "pub fn increase_count(mut self, step: int) {\n\t\tset self.count += step;\n\t}",
                                start: 184,
                                end: 257,
                            },
                            inner: [
                                Pair {
                                    rule: pub,
                                    span: Span {
                                        str: "pub",
                                        start: 184,
                                        end: 187,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: fn,
                                    span: Span {
                                        str: "fn",
                                        start: 188,
                                        end: 190,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "increase_count",
                                        start: 191,
                                        end: 205,
                                    },
                                    inner: [],
                                },
                                Pair {
                                    rule: impl_function_parameters_list,
                                    span: Span {
                                        str: "mut self, step: int",
                                        start: 206,
                                        end: 225,
                                    },
                                    inner: [
                                        Pair {
                                            rule: mutable,
                                            span: Span {
                                                str: "mut",
                                                start: 206,
                                                end: 209,
                                            },
                                            inner: [],
                                        },
                                        Pair {
                                            rule: self_param,
                                            span: Span {
                                                str: "self",
                                                start: 210,
                                                end: 214,
                                            },
                                            inner: [],
                                        },
                                        Pair {
                                            rule: function_parameter,
                                            span: Span {
                                                str: "step: int",
                                                start: 216,
                                                end: 225,
                                            },
                                            inner: [
                                                Pair {
                                                    rule: ident,
                                                    span: Span {
                                                        str: "step",
                                                        start: 216,
                                                        end: 220,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: type,
                                                    span: Span {
                                                        str: "int",
                                                        start: 222,
                                                        end: 225,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: ident,
                                                            span: Span {
                                                                str: "int",
                                                                start: 222,
                                                                end: 225,
                                                            },
                                                            inner: [],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                },
                                Pair {
                                    rule: block_expr,
                                    span: Span {
                                        str: "{\n\t\tset self.count += step;\n\t}",
                                        start: 227,
                                        end: 257,
                                    },
                                    inner: [
                                        Pair {
                                            rule: mutate_stmt,
                                            span: Span {
                                                str: "set self.count += step;",
                                                start: 231,
                                                end: 254,
                                            },
                                            inner: [
                                                Pair {
                                                    rule: set,
                                                    span: Span {
                                                        str: "set",
                                                        start: 231,
                                                        end: 234,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: assignment_accessor,
                                                    span: Span {
                                                        str: "self.count",
                                                        start: 235,
                                                        end: 245,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: increment_assign,
                                                    span: Span {
                                                        str: "+=",
                                                        start: 246,
                                                        end: 248,
                                                    },
                                                    inner: [],
                                                },
                                                Pair {
                                                    rule: expr,
                                                    span: Span {
                                                        str: "step",
                                                        start: 249,
                                                        end: 253,
                                                    },
                                                    inner: [
                                                        Pair {
                                                            rule: term,
                                                            span: Span {
                                                                str: "step",
                                                                start: 249,
                                                                end: 253,
                                                            },
                                                            inner: [
                                                                Pair {
                                                                    rule: ident,
                                                                    span: Span {
                                                                        str: "step",
                                                                        start: 249,
                                                                        end: 253,
                                                                    },
                                                                    inner: [],
                                                                },
                                                            ],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
            ],
        },
        Pair {
            rule: EOI,
            span: Span {
                str: "",
                start: 261,
                end: 261,
            },
            inner: [],
        },
    ],
}
